cmd/thanos/endpointset.go

```diff
@@ -0,0 +1,350 @@
+	staticStrictEndpointGroups []string,
+) (*endpointConfigProvider, error) {
+	res := &endpointConfigProvider{
+		endpoints:            staticEndpoints,
+		endpointGroups:       staticEndpointGroups,
+		strictEndpoints:      staticStrictEndpoints,
+		strictEndpointGroups: staticStrictEndpointGroups,
+	}
+
+	if configFile == nil {
+		configFile = extkingpin.NewNopConfig()
+	}
+
+	cfg, err := res.parse(configFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load config file")
+	}
+	res.addStaticEndpoints(&cfg)
+	res.cfg = cfg
+	if err := validateEndpointConfig(cfg); err != nil {
+		return nil, errors.Wrapf(err, "unable to validate endpoints")
+	}
+
+	// only static endpoints
+	if len(configFile.Path()) == 0 {
+		return res, nil
+	}
+
+	if err := extkingpin.PathContentReloader(context.Background(), configFile, logger, func() {
+		res.mu.Lock()
+		defer res.mu.Unlock()
+
+		level.Info(logger).Log("msg", "reloading endpoint config")
+		cfg, err := res.parse(configFile)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to reload endpoint config", "err", err)
+			return
+		}
+		res.addStaticEndpoints(&cfg)
+		if err := validateEndpointConfig(cfg); err != nil {
+			level.Error(logger).Log("msg", "failed to validate endpoint config", "err", err)
+			return
+		}
+		res.cfg = cfg
+	}, configReloadInterval); err != nil {
+		return nil, errors.Wrapf(err, "unable to create config reloader")
+	}
+	return res, nil
+}
+
+func setupEndpointSet(
+	g *run.Group,
+	comp component.Component,
+	reg prometheus.Registerer,
+	logger log.Logger,
+	configFile fileContent,
+	configReloadInterval time.Duration,
+	legacyFileSDFiles []string,
+	legacyFileSDInterval time.Duration,
+	legacyEndpoints []string,
+	legacyEndpointGroups []string,
+	legacyStrictEndpoints []string,
+	legacyStrictEndpointGroups []string,
+	dnsSDResolver string,
+	dnsSDInterval time.Duration,
+	unhealthyTimeout time.Duration,
+	endpointTimeout time.Duration,
+	dialOpts []grpc.DialOption,
+	queryConnMetricLabels ...string,
+) (*query.EndpointSet, error) {
+	configProvider, err := newEndpointConfigProvider(
+		logger,
+		configFile,
+		configReloadInterval,
+		legacyEndpoints,
+		legacyEndpointGroups,
+		legacyStrictEndpoints,
+		legacyStrictEndpointGroups,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load config initially")
+	}
+	// Register resolver for the "thanos:///" scheme for endpoint-groups
+	dns.RegisterGRPCResolver(
+		logger,
+		dns.NewProvider(
+			logger,
+			extprom.WrapRegistererWithPrefix(fmt.Sprintf("thanos_%s_endpoint_groups_", comp), reg),
+			dns.ResolverType(dnsSDResolver),
+		),
+		dnsSDInterval,
+	)
+	dnsEndpointProvider := dns.NewProvider(
+		logger,
+		extprom.WrapRegistererWithPrefix(fmt.Sprintf("thanos_%s_endpoints_", comp), reg),
+		dns.ResolverType(dnsSDResolver),
+	)
+	duplicatedStores := promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: fmt.Sprintf("thanos_%s_duplicated_store_addresses_total", comp),
```

Comment:
Can we use the word "endpoint" here instead of "store"? In the variable as well :pray:

Reply:
done!